.bulldozer.yml

```diff
@@ -0,0 +1,66 @@
+# "version" is the configuration version, currently "1".
+version: 1
+
+# "merge" defines how and when pull requests are merged. If the section is
+# missing, bulldozer will consider all pull requests and use default settings.
+merge:
+  # "whitelist" defines the set of pull requests considered by bulldozer. If
+  # the section is missing, bulldozer considers all pull requests not excluded
+  # by the blacklist.
+  whitelist:
+    # Pull requests with any of these labels (case-insensitive) are added to
+    # the whitelist.
+    labels: ["merge when ready"]
+
+  # "blacklist" defines the set of pull request ignored by bulldozer. If the
+  # section is missing, bulldozer considers all pull requests. It takes the
+  # same keys as the "whitelist" section.
+  blacklist:
+    labels: ["do not merge"]
+    comment_substrings: ["==DO_NOT_MERGE==", "WIP"]
+
+  # "method" defines the merge method. The available options are "merge",
+  # "rebase", and "squash".
+  method: squash
+
+  # Allows the merge method that is used when auto-merging a PR to be different based on the
+  # target branch. The keys of the hash are the target branch name, and the values are the merge method that
+  # will be used for PRs targeting that branch. The valid values are the same as for the "method" key.
+  # Note: If the target branch does not match any of the specified keys, the "method" key is used instead.
+  branch_method:
+    master: squash
+
+  # "options" defines additional options for the individual merge methods.
+  options:
+    # "squash" options are only used when the merge method is "squash"
+    squash:
+      # "body" defines how the body of the commit message is created when
+      # generating a squash commit. The options are "pull_request_body",
+      # "summarize_commits", and "empty_body".
+      body: "pull_request_body"
+      # if "body" is "pull_request_body" then the commit message will be part
+      # of the pull request body delinated by "message_delimiter" string
+      message_delimiter: ==COMMIT_MSG==
+
+  # "required_status" is a list of additional status contexts that must pass
+  # before bulldozer can merge a pull request. This is useful if you want to
+  # require extra testing for automated merges, but not for manual merges.
+  #required_statuses:
+  #  - "ci/circleci: ete-tests"
+
+  # If true, bulldozer will delete branches after their pull requests merge.
+  delete_after_merge: true
+
+## "update" defines how and when to update pull request branches. Unlike with
+## merges, if this section is missing, bulldozer will not update any pull requests.
+#update:
```

Comment:
Assuming tests are run as part of PR approval and required before merge, this shouldn't arise (i.e. step 2 checks fail). Though not having to manually update branches is definitely a useful feature of bulldozer and worth using.

> I think it's cool to let the software update the branch before merging. Otherwise, the following could happen:
> 
> 1. There are branches A and B, both with test passing.
> 2. Branch A gets merged into master. Everything is good in master.
> 3. Branch B gets merged into master. However, when integrating the changes from A, which cause no conflict, tests stop working for some reason, some incompatibility between changes. Then, master is broken.
> 
> The drawback is that if there's a bunch of PRs ready to be merged, they are gonna update a lot and thus the CI is gonna be used a lot.

Reply:
We've explicitly avoided doing this when merging manually because, while the issue you pointed out is definitely possible, it's rare. The additional CI delay, unfortunately, is then always incurred. With bulldozer though I'd be less worried because it's just machine time, not developer time.

The real solution to this issue would be to use a build queue that takes the next PR off of the queue, merges in master, runs CI, and merges the PR into master if everything passed. Only the build queue would merge to master and it would do so synchronously, so the issue you highlight would be avoided. I haven't researched solutions for this, but if you know of any, please let us know!

Reply:
@viveklak, I believe @bryant1410 is postulating a situation where change A + master passes, B + master passes, but A + B + master fails. Since we don't require that master be merged back into B after A is merged to master, this could occur. Please correct me if I'm misunderstanding something about bulldozer!

Reply:
> We've explicitly avoided doing this when merging manually because, while the issue you pointed out is definitely possible, it's rare. The additional CI delay, unfortunately, is then always incurred. With bulldozer though I'd be less worried because it's just machine time, not developer time.

Yeah, it's unusual.

> The real solution to this issue would be to use a build queue that takes the next PR off of the queue, merges in master, runs CI, and merges the PR into master if everything passed. Only the build queue would merge to master and it would do so synchronously, so the issue you highlight would be avoided. I haven't researched solutions for this, but if you know of any, please let us know!

Isn't it the same? Cause if bulldozer merges master into the branch so it's up-to-date then it's gonna have the same result as if the branch is merged into master (or rebased), right? (and then it runs the tests). Or am I missing something?

I've recently read [Shopify's solution to this](https://engineering.shopify.com/blogs/engineering/introducing-the-merge-queue). And saw [a guy's response that he/they chose mergify](https://twitter.com/santib6_/status/1217970721634365443).

Reply:
I don't know much about bulldozer, but my assumption is that it doesn't actually have a queue, so we would end up running CI for PR B that would be invalidated after PR A was merged. They're equally safe, but a queue would use less CI time/resources.

Reply:
And thanks for the links!!

Reply:
> @viveklak, I believe @bryant1410 is postulating a situation where change A + master passes, B + master passes, but A + B + master fails. Since we don't require that master be merged back into B after A is merged to master, this could occur. Please correct me if I'm misunderstanding something about bulldozer!

Ah - I have typically run with the following setting turned on  (`Require branches to be up to date before merging`):
<img width="691" alt="Screen Shot 2020-01-16 at 6 17 38 PM" src="https://user-images.githubusercontent.com/1094518/72578981-92576780-388c-11ea-9050-3232870bbd8c.png">

We use circleci which supports auto-cancel for redundant builds on branches running off previous revisions which helps trim the potential splay of CI builds.

Reply:
Yep, that's precisely the setting we've explicitly avoided. ;)

Reply:
> I don't know much about bulldozer, but my assumption is that it doesn't actually have a queue, so we would end up running CI for PR B that would be invalidated after PR A was merged. They're equally safe, but a queue would use less CI time/resources.

Oh, yeah, I was supposing the CI can handle one run at a time :smile:

Reply:
> Ah - I have typically run with the following setting turned on (`Require branches to be up to date before merging`):
> <img alt="Screen Shot 2020-01-16 at 6 17 38 PM" width="691" src="https://user-images.githubusercontent.com/1094518/72578981-92576780-388c-11ea-9050-3232870bbd8c.png">
> 
> We use circleci which supports auto-cancel for redundant builds on branches running off previous revisions which helps trim the potential splay of CI builds.

Yeah, that seems like the best to avoid 2 passing branches to be merged, and also the auto-cancel thing.

However, how do you avoid 2 PRs of being checked at the same time? If this restrictions are on, one of them is gonna be merged and then the other one canceled and run again.

Reply:
> I don't know much about bulldozer, but my assumption is that it doesn't actually have a queue, so we would end up running CI for PR B that would be invalidated after PR A was merged. They're equally safe, but a queue would use less CI time/resources.

Indeed. All bulldozer is doing is listening for webhooks for PRs with the specified whitelist, attempting to update/merge where the github repo configuration permits. The stated goal is to reduce devs babysiting merges when CI runs take a long time or if the repo gets updated a bunch. Some of the solutions that @bryant1410 mentioned are probably a lot more powerful.

Reply:
> Yep, that's precisely the setting we've explicitly avoided. ;)

Or rather, I thought we had... @schmmd, did you reenable "Require branches to be up to date before merging"?

Reply:
@brendan-ai2, @matt-gardner re-enabled it.  Unfortunately when it's disabled the "merge in master" button disappears altogether and that button is pretty useful.  We mentioned this in Slack--but it would have been easy to miss.